File: BOJ/Gold I/BOJ2042.py
#BOJ2042 구간 합 구하기 20210425
import sys
from math import log2,ceil
input = sys.stdin.readline

def makeSegmentTree(node, start, end, li, tree):
    if start == end:
        tree[node] = li[start]
        return tree[node]
    else:
        left = makeSegmentTree(node*2, start, (start+end)//2, li, tree)
        right = makeSegmentTree((node*2)+1, (start+end)//2 + 1, end, li, tree)
        tree[node] = left + right
        return tree[node]

def subSum(node, start, end, l, r, tree):
    if r < start or l > end:
        return 0
    if l <= start and end <= r:
        return tree[node]
    left = subSum(node*2, start, (start+end)//2, l, r, tree)
    right = subSum(node*2+1, (start+end)//2 + 1 , end, l, r, tree)
    return left + right

def update(node, start, end, idx, diff, tree):
    if not(start <= idx <= end):
        return 
    tree[node] += diff
    if start != end:
        update(node*2, start, (start+end)//2, idx, diff, tree)
        update(node*2+1, (start+end)//2 + 1 , end, idx, diff, tree)
        
UPDATE = 1
SUBSUM = 2
def query(what, a, b, tree, n, li):
    if what == UPDATE:
        diff = b - li[a] + 1
        li[a] = b + 1
        update(1, 0, n-1, a, diff, tree)
    elif what == SUBSUM:
        print(subSum(1, 0, n-1, a, b, tree))

def main():
    n,m,k = map(int,input().rstrip().split())
    li = [int(input()) for _ in range(n)]
    tree = [0]+[0] * (2**(ceil(log2(n))+1))
    makeSegmentTree(1,0,n-1,li,tree)
    
    for _ in range(m+k):
        what, a, b = map(int,input().rstrip().split())
        query(what, a-1, b-1, tree, n, li)
if __name__ == '__main__':
    main()

File: BOJ/Gold I/BOJ2042.java
//BOJ2042 구간 합 구하기 20210708
import java.io.*;
import java.util.*;
import java.util.stream.Collectors;

import static java.lang.Math.*;

public class Main {
    private static final BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static long[] arr;
    private static long[] tree;

    public static void main(String[] args) throws IOException {
        StringTokenizer stk = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(stk.nextToken());
        int m = Integer.parseInt(stk.nextToken());
        int k = Integer.parseInt(stk.nextToken());
        arr = new long[n];
        for(int i = 0; i < n; i++){
            arr[i] = Long.parseLong(br.readLine());
        }

        tree = new long[(int) pow(2, ceil(log2(n)) + 1)];
        makeSegmentTree(1, 0, n-1);

        for(int i = 0; i < m + k; i++){
            String[] cmd = br.readLine().split(" ");
            int a = Integer.parseInt(cmd[0]), b = Integer.parseInt(cmd[1]) - 1;
            if(a == 1){
                long c = Long.parseLong(cmd[2]);
                update(1, 0, n-1, b, arr[b], c);
                arr[b] = c;
            }else{
                int c = Integer.parseInt(cmd[2]);
                bw.write(getSum(1, 0, n-1, b, c-1) + "\n");
            }
        }
        bw.flush();
        bw.close();
    }

    public static double log2(int n){
        return log10(n)/log10(2);
    }

    public static long makeSegmentTree(int node, int start, int end){
        if(start == end){
            tree[node] = arr[start];
            return tree[node];
        }
        int mid = (start + end) / 2;
        long left = makeSegmentTree(node * 2, start, mid);
        long right = makeSegmentTree(node * 2 + 1, mid + 1, end);
        tree[node] = left + right;
        return tree[node];
    }

    public static void update(int node, int start, int end, int idx, long before, long after){
        if(idx < start || end < idx) return;
        if(start == end){
            tree[node] = after;
            return;
        }
        tree[node] = tree[node] - before + after;
        int mid = (start + end) / 2;
        update(node * 2, start, mid, idx, before, after);
        update(node * 2 + 1, mid+1, end, idx, before, after);
    }

    public static long getSum(int node, int start, int end, int left, int right){
        if(right < start || end < left) return 0L;
        if(left <= start && end <= right){
            return tree[node];
        }
        int mid = (start + end) / 2;
        long l = getSum(node * 2, start, mid, left, right);
        long r = getSum(node * 2 + 1, mid + 1, end, left, right);
        return l + r;
    }
}